walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,94 @@
+    
+    // If there are no required attendees, then the meeting can be held at any time of the day.
+    if (meetingAttendees.isEmpty()) {
+      meetingTimes.add(TimeRange.WHOLE_DAY);
+      return meetingTimes;
+    }
+
+    // If the meeting lasts longer than the day, then there are no meeting time slots available.
+    if (meetingDuration > TimeRange.WHOLE_DAY.duration()) {
+      return meetingTimes;
+    }
+
+    // Organize events by start time in increasing order.
+    List<Event> eventList = new ArrayList<>(events);
+    Collections.sort(eventList, new Comparator<Event>() {
+      public int compare(Event event1, Event event2) {
+        return Integer.compare(event1.getWhen().start(), event2.getWhen().start());
+      }
+    });
+
+    // Remove all events that do not include the meeting attendees.
+    eventList.removeIf(event ->
+        !(new HashSet<>(event.getAttendees()).removeAll(request.getAttendees())));
+
+    // Start with an open schedule and split it up based on events.
+    meetingTimes.add(TimeRange.WHOLE_DAY);
+    for (Event event: eventList) {
+      meetingTimes = splitSchedule(meetingTimes, event);
+    }
+
+    // Remove all time slots that are not long enough to hold the meeting.
+    meetingTimes.removeIf(time -> time.duration() < request.getDuration());
+    return meetingTimes;
+  }
+
+  /** 
+   * Splits up the schedule to accomodate a particular event.
+   *
+   * @param meetingTimes The current available meeting time slots.
+   * @param event The event that you are trying to build meeting times around.
+   *
+   * @return The new available meeting time slots after accomodating the event.
+   */
+  private Collection<TimeRange> splitSchedule(Collection<TimeRange> meetingTimes, Event event) {
+    int eventStart = event.getWhen().start();
+    int eventEnd = event.getWhen().end();
+    Collection<TimeRange> newMeetingTimes = new ArrayList();
+    for (TimeRange time: meetingTimes) {
+      int timeSlotStart = time.start();
+      int timeSlotEnd = time.end();
+      if (timeSlotStart < eventStart && timeSlotEnd > eventEnd) {
+        // Handles the case of when an event is entirely within a time slot.
+        newMeetingTimes.add(TimeRange.fromStartEnd(timeSlotStart, eventStart, false));
+        newMeetingTimes.add(TimeRange.fromStartEnd(eventEnd, timeSlotEnd, false));
+
+      } else if (timeSlotStart < eventStart && timeSlotEnd > eventStart) {
+        // Handles the case of when an event starts within the time slot,
+        // but ends after the time slots end time.
+        newMeetingTimes.add(TimeRange.fromStartEnd(timeSlotStart, eventStart, false));
+
+      } else if (timeSlotStart < eventEnd && timeSlotEnd > eventEnd) {
+        // Handles the case of when an event ends after the time slot,
+        // but starts before the time slot ends.
+        newMeetingTimes.add(TimeRange.fromStartEnd(eventEnd, timeSlotEnd, false));
+
+      } else if (timeSlotStart == eventStart && timeSlotEnd > eventEnd) {
```

Comment:
can you use <= or >= to combine these cases with some of the ones above? if not, add a comment explaining why they are necessary. there are a lot of different cases here